src/client/coll.rs

```diff
@@ -0,0 +1,201 @@
+        let mut new_opts = self.clone();
+        new_opts.skip = skip;
+        new_opts
+    }
+
+    pub fn with_limit(&self, limit: i32) -> FindOptions {
+        let mut new_opts = self.clone();
+        new_opts.limit = limit;
+        new_opts
+    }
+
+    pub fn with_projection(&self, proj: Option<bson::Document>) -> FindOptions {
+        let mut new_opts = self.clone();
+        new_opts.projection = proj;
+        new_opts
+    }
+}
+
+impl<'a> Collection<'a> {
+    /// Creates a collection representation with optional read and write controls.
+    ///
+    /// If `create` is specified, the collection will be explicitly created in the database.
+    pub fn new(db: &'a Database<'a>, name: &str, create: bool,
+               read_preference: Option<ReadPreference>, write_concern: Option<WriteConcern>) -> Collection<'a> {
+
+        let coll = Collection {
+            full_name: format!("{}.{}", db.name, name),
+            db: db,
+            name: name.to_owned(),
+            read_preference: read_preference,
+            write_concern: write_concern,
+        };
+
+        /*
+        // Since standard collections are implicitly created on insert,
+        // this should only be used to create capped collections.
+        if create {
+            coll.create();
+        };
+         */
+
+        coll
+    }
+
+    /// Returns a unique operational request id.
+    pub fn get_req_id(&self) -> i32 {
+        self.db.client.get_req_id()
+    }
+
+    // Read Spec
+    pub fn aggregate(pipeline: &[bson::Document], options: AggregateOptions) -> Result<Vec<bson::Document>, String> {
+        Err("IMPL".to_owned())
+    }
+
+    pub fn count(filter: bson::Document, options: CountOptions) -> Result<i64, String> {
+        Err("IMPL".to_owned())
+    }
+
+    pub fn distinct(field_name: &str, filter: bson::Document, options: DistinctOptions) -> Result<Vec<String>, String> {
+        Err("IMPL".to_owned())
+    }
+
+    /// Returns a list of documents within the collection that match the filter.
+    pub fn find(&self, filter: Option<bson::Document>, options: Option<FindOptions>)
+                -> Result<Vec<bson::Document>, String> {
+
+        let doc = match filter {
+            Some(bson) => bson,
+            None => bson::Document::new(),
+        };
+
+        let options = match options {
+            Some(opts) => opts,
+            None => FindOptions::new(),
+        };
+
+        let flags = OpQueryFlags {
+            tailable_cursor: options.cursor_type != CursorType::NonTailable,
+            slave_ok: false,
+            oplog_relay: options.op_log_replay,
+            no_cursor_timeout: options.no_cursor_timeout,
+            await_data: options.cursor_type == CursorType::TailableAwait,
+            exhaust: false,
+            partial: options.allow_partial_results,
+        };
+
+        let req = try!(Message::with_query(self.get_req_id(), flags, self.full_name.to_owned(),
```

Comment:
In general, we should probably abstract out the wire protocol a little more here. Generally find() shouldn't interact with a socket directly.